File: leetcode621/leetcode621_2.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        vector<int> counts(26);
        
        for (auto t : tasks) {
            ++counts[t - 'A'];
        }
        
        priority_queue<pair<int, char>> pq;
        for (int i = 0; i < 26; ++i) {
            if (counts[i] > 0) {
                pq.push({counts[i], i + 'A'});
            }
        }
        
        int ans = 0;
        int cycle = n + 1;
        
        // one cycle
        while (!pq.empty()) {
            int time = 0;
            vector<pair<int, char>> tmp;
            
            // assign tasks
            for (int i = 0; i < cycle; ++i) {
                // do this task
                if (!pq.empty()) {
                    tmp.push_back(pq.top());
                    pq.pop();
                    tmp.back().first--;
                    ++time;
                }
                else {
                    break;
                }
            }
            
            // put back for next cycle
            for (auto t : tmp) {
                if (t.first > 0) {
                    pq.push(t);
                }
            }
            
            // assign idles? last cycle?
            ans += pq.empty() ? time : cycle;
        }
        
        return ans;
    }
};

// Runtime: 140 ms, faster than 30.77% of C++ online submissions for Task Scheduler.
// Memory Usage: 24.3 MB, less than 6.38% of C++ online submissions for Task Scheduler.
// time complexity: O(n + n * m), m is the cycle time
// space complexity: O(n)
// https://leetcode.com/problems/task-scheduler/discuss/104493/C%2B%2B-Java-Clean-Code-Priority-Queue

// class Solution {
// public:
//     int leastInterval(vector<char>& tasks, int n) {
//         unordered_map<char, int> counts;
//         for (char t : tasks) {
//             counts[t]++;
//         }
//         priority_queue<pair<int, char>> pq;
//         for (pair<char, int> count : counts) {
//             pq.push(make_pair(count.second, count.first));
//         }
//         int alltime = 0;
//         int cycle = n + 1;
//         while (!pq.empty()) {
//             int time = 0;
//             vector<pair<int, char>> tmp;
//             for (int i = 0; i < cycle; i++) {
//                 if (!pq.empty()) {
//                     tmp.push_back(pq.top());
//                     pq.pop();
//                     time++;
//                 }
//             }
//             for (auto t : tmp) {
//                 if (--t.first) {
//                     pq.push(t);
//                 }
//             }
//             alltime += !pq.empty() ? cycle : time;
//         }
//         return alltime;
//     }
// };

File: leetcode621/leetcode621_2.java
public class Solution {
    public int leastInterval(char[] tasks, int n) {
        Map<Character, Integer> counts = new HashMap<Character, Integer>();
        for (char t : tasks) {
            counts.put(t, counts.getOrDefault(t, 0) + 1);
        }

        PriorityQueue<Integer> pq = new PriorityQueue<Integer>((a, b) -> b - a);
        pq.addAll(counts.values());

        int alltime = 0;
        int cycle = n + 1;

        while (!pq.isEmpty()) {
            int worktime = 0;
            List<Integer> tmp = new ArrayList<Integer>();

            for (int i = 0; i < cycle; i++) {
                if (!pq.isEmpty()) {
                    tmp.add(pq.poll());
                    worktime++;
                }
            }

            for (int cnt : tmp) {
                if (cnt - 1 > 0) {
                    pq.offer(cnt - 1);
                }
            }

            alltime += !pq.isEmpty() ? cycle : worktime;
        }

        return alltime;
    }
}

// Runtime: 24 ms, faster than 30.45% of Java online submissions for Task
// Scheduler.
// Memory Usage: 40.3 MB, less than 54.27% of Java online submissions for Task
// Scheduler.
// time complexity: O(n log n)
// space complexity: O(n)
// https://leetcode.com/problems/task-scheduler/discuss/104493/C%2B%2B-Java-Clean-Code-Priority-Queue